### PR TITLE
expression: prevent function of castYearAsTime from pushing down to Ti… (#25975)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #25975 to release-5.0
You can switch your code base to this Pull Request by using [git-extras](https://github.com/tj/git-extras):
```bash
# In tidb repo:
git pr https://github.com/pingcap/tidb/pull/26369
```

After apply modifications, you can push your change to this PR via:
```bash
git push git@github.com:ti-srebot/tidb.git pr/26369:release-5.0-39a56e5c4df0
```

---

### What problem does this PR solve?

Issue Number: https://github.com/pingcap/tics/issues/1964 

Problem Summary:

The type of year operation with the type of date is wrong, for example:
```sql
create table t(a year,b date);
insert into t values(2005,"2004-10-10"),(2015,"2018-03-27");
alter table t set tiflash replica 1;
select * from t where a>b;
```
except:
```sql
+------+------------+
| a    | b          |
+------+------------+
| 2005 | 2004-10-10 |
+------+------------+
```
actual:
```sql
empty
```
### What is changed and how it works?
the problem is because that the column `a` of type year handled by function of castIntAsTime is pushed down to Tiflash.
Tiflash cannot recognize type year without field type,  so cast the column a as time incorrectly.
Temporarily stop pushing down the function of castIntAsTime to TiFlash to get the correctly record.

### Release note
* no release note

### CODE REVIEW

File: expression/expression.go [hunk 1]
```diff
@@ -1018,11 +1018,22 @@ func scalarExprSupportedByFlash(function *ScalarFunction) bool {
 		}
 	case ast.Cast:
 		switch function.Function.PbCode() {
+<<<<<<< HEAD
 		case tipb.ScalarFuncSig_CastIntAsInt, tipb.ScalarFuncSig_CastIntAsDecimal, tipb.ScalarFuncSig_CastIntAsString, tipb.ScalarFuncSig_CastIntAsTime,
 			tipb.ScalarFuncSig_CastRealAsInt, tipb.ScalarFuncSig_CastRealAsDecimal, tipb.ScalarFuncSig_CastRealAsString, tipb.ScalarFuncSig_CastRealAsTime,
 			tipb.ScalarFuncSig_CastStringAsInt, tipb.ScalarFuncSig_CastStringAsDecimal, tipb.ScalarFuncSig_CastStringAsString, tipb.ScalarFuncSig_CastStringAsTime, tipb.ScalarFuncSig_CastStringAsReal,
 			tipb.ScalarFuncSig_CastDecimalAsInt, tipb.ScalarFuncSig_CastDecimalAsDecimal, tipb.ScalarFuncSig_CastDecimalAsString, tipb.ScalarFuncSig_CastDecimalAsTime,
 			tipb.ScalarFuncSig_CastTimeAsInt, tipb.ScalarFuncSig_CastTimeAsDecimal, tipb.ScalarFuncSig_CastTimeAsString, tipb.ScalarFuncSig_CastTimeAsTime:
+=======
```
Comment (zimulala) [inline expression/expression.go:1027]: Please fix the conflict @mengxin9014

File: expression/expression.go [hunk 2]
```diff
@@ -1018,11 +1018,22 @@ func scalarExprSupportedByFlash(function *ScalarFunction) bool {
 		}
 	case ast.Cast:
 		switch function.Function.PbCode() {
+<<<<<<< HEAD
```
Comment (mengxin9014) [inline expression/expression.go:1021]:
```suggestion
```

File: expression/expression.go [hunk 3]
```diff
@@ -1018,11 +1018,22 @@ func scalarExprSupportedByFlash(function *ScalarFunction) bool {
...
 		case tipb.ScalarFuncSig_CastIntAsInt, tipb.ScalarFuncSig_CastIntAsDecimal, tipb.ScalarFuncSig_CastIntAsString, tipb.ScalarFuncSig_CastIntAsTime,
```
Comment (mengxin9014) [inline expression/expression.go:1022]:
```suggestion
```

File: expression/expression.go [hunk 4]
```diff
@@ -1018,11 +1018,22 @@ func scalarExprSupportedByFlash(function *ScalarFunction) bool {
...
 			tipb.ScalarFuncSig_CastRealAsInt, tipb.ScalarFuncSig_CastRealAsDecimal, tipb.ScalarFuncSig_CastRealAsString, tipb.ScalarFuncSig_CastRealAsTime,
```
Comment (mengxin9014) [inline expression/expression.go:1023]:
```suggestion
```

File: expression/expression.go [hunk 5]
```diff
@@ -1018,11 +1018,22 @@ func scalarExprSupportedByFlash(function *ScalarFunction) bool {
...
 			tipb.ScalarFuncSig_CastStringAsInt, tipb.ScalarFuncSig_CastStringAsDecimal, tipb.ScalarFuncSig_CastStringAsString, tipb.ScalarFuncSig_CastStringAsTime, tipb.ScalarFuncSig_CastStringAsReal,
```
Comment (mengxin9014) [inline expression/expression.go:1024]:
```suggestion
```

File: expression/expression.go [hunk 6]
```diff
@@ -1018,11 +1018,22 @@ func scalarExprSupportedByFlash(function *ScalarFunction) bool {
...
 			tipb.ScalarFuncSig_CastDecimalAsInt, tipb.ScalarFuncSig_CastDecimalAsDecimal, tipb.ScalarFuncSig_CastDecimalAsString, tipb.ScalarFuncSig_CastDecimalAsTime,
```
Comment (mengxin9014) [inline expression/expression.go:1025]:
```suggestion
```

File: expression/expression.go [hunk 7]
```diff
@@ -1018,11 +1018,22 @@ func scalarExprSupportedByFlash(function *ScalarFunction) bool {
...
 			tipb.ScalarFuncSig_CastTimeAsInt, tipb.ScalarFuncSig_CastTimeAsDecimal, tipb.ScalarFuncSig_CastTimeAsString, tipb.ScalarFuncSig_CastTimeAsTime:
```
Comment (mengxin9014) [inline expression/expression.go:1026]:
```suggestion
```

File: expression/expression.go (hunk 1 above)
Comment (mengxin9014) [reply expression/expression.go:1027]:
```suggestion
```

**Review by LittleFall — APPROVED** [review]
LGTM

**Review by zanmato1984 — APPROVED** [review]
LGTM

**Pull request merged**